kroxylicious-additional-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/EnvelopeEncryptionFilter.java

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.filter.encryption;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+import org.apache.kafka.common.message.FetchResponseData;
+import org.apache.kafka.common.message.FetchResponseData.FetchableTopicResponse;
+import org.apache.kafka.common.message.FetchResponseData.PartitionData;
+import org.apache.kafka.common.message.ProduceRequestData;
+import org.apache.kafka.common.message.ProduceRequestData.PartitionProduceData;
+import org.apache.kafka.common.message.ProduceRequestData.TopicProduceData;
+import org.apache.kafka.common.message.RequestHeaderData;
+import org.apache.kafka.common.message.ResponseHeaderData;
+import org.apache.kafka.common.record.MemoryRecords;
+import org.apache.kafka.common.record.MemoryRecordsBuilder;
+import org.apache.kafka.common.record.RecordBatch;
+
+import io.kroxylicious.proxy.filter.FetchResponseFilter;
+import io.kroxylicious.proxy.filter.FilterContext;
+import io.kroxylicious.proxy.filter.ProduceRequestFilter;
+import io.kroxylicious.proxy.filter.RequestFilterResult;
+import io.kroxylicious.proxy.filter.ResponseFilterResult;
+
+/**
+ * A filter for encrypting and decrypting records using envelope encryption
+ * @param <K> The type of KEK reference
+ * @param <E> The type of wrapped DEK
+ */
+class EnvelopeEncryptionFilter<K, E>
+        implements ProduceRequestFilter, FetchResponseFilter {
+    private final TopicNameBasedKekSelector<K> kekSelector;
+
+    private final DekCache<K, E> dekCache;
+
+    private BufferPool bufferPool;
+
+    EnvelopeEncryptionFilter(DekCache<K, E> dekCache, TopicNameBasedKekSelector<K> kekSelector) {
+        this.kekSelector = kekSelector;
+        this.dekCache = dekCache;
+    }
+
+    @SuppressWarnings("unchecked")
+    static <T> CompletableFuture<List<T>> join(List<CompletableFuture<T>> futures) {
+        return CompletableFuture.allOf(futures.toArray((CompletableFuture<T>[]) new CompletableFuture[futures.size()])).thenApply(ignored -> {
+            return futures.stream().map(cf -> {
+                try {
+                    return cf.get();
```

Comment:
Eliminates the need for the exception handling
```suggestion
                    return cf.join();
```

Reply:
Tom addressed this.